### DISCUSSION TITLE
Import structs from a library

### DISCUSSION THREAD

**Discussion by Leastrio** [rusterlium/rustler#576] [Q&A]
Is there a way to import structs from another library in rust to elixir without having to rewrite the entire struct again? Been looking through the docs and cant find much on this.

**Comment by filmor** [comment]
This is a general (intentional) limitation of Rust's trait system: To implement a trait for a type in a crate, you have to either implement the trait or the type in the same crate. In this case, you have a type from a foreign crate `SomeType` and the traits come from Rustler (`rustler::Encoder` and `rustler::Decoder`).

The "common" approach to this for "non-generated" trait implementations is to wrap the type in a new-type (a tuple `struct` with a single element, `pub struct NewSomeType(SomeType);`), accompany it with a `impl From<SomeType> for NewSomeType` and implement the required trait on it.

That won't help you here, as the derive macros that provide the implementations for `Encoder` and `Decoder` need to "see" the shape of the struct to convert on the syntax-level.

The best I can think of right now is to implement `Encoder` and `Decoder` manually on a wrapped newtype.

**Reply by Leastrio** [reply]
Hmm okay im not sure if thats viable as im looking to convert a LOT of types over from this one library to elixir but maybe I need to look at this in another way

**Comment by sachanganesh** [comment]
@filmor  Thank you for the thorough explanation!

I know the discussion may be stale, but I was wondering if you could provide a simple example of how this manual implementation of Encoder and Decoder would be done.

```rust
// external struct 
pub struct TestStruct {
    pub year: usize,
    pub month: usize,
}

struct Wrapper(TestStruct);

impl Encoder for DateTime {
    fn encode<'b>(&self, env: Env<'b>) -> Term<'b> {
        // I doubt this is correct

        match elixir_struct::make_ex_struct(env, "ICalendar.DateTime") {
            Ok(term) => term,
            Err(err) => panic!("{:#?}", err),
        }
    }
}

impl<'a> Decoder<'a> for DateTime {
    fn decode(term: Term<'a>) -> NifResult<Self> {
        // unsure
    }
}
```
